src/engine/sim_engine/pool.rs

```diff
@@ -476,14 +476,33 @@ impl Pool for SimPool {
         &mut self,
         _pool_name: &str,
         fs_uuids: &[FilesystemUuid],
-    ) -> StratisResult<SetDeleteAction<FilesystemUuid>> {
+    ) -> StratisResult<SetDeleteAction<FilesystemUuid, FilesystemUuid>> {
         let mut removed = Vec::new();
         for &uuid in fs_uuids {
             if self.filesystems.remove_by_uuid(uuid).is_some() {
                 removed.push(uuid);
             }
         }
-        Ok(SetDeleteAction::new(removed))
+
+        let snapshots: Vec<FilesystemUuid> = self
+            .filesystems()
+            .iter()
+            .filter_map(|(_, u, fs)| {
+                fs.origin()
+                    .and_then(|x| if removed.contains(&x) { Some(*u) } else { None })
+            })
+            .collect();
+
+        let mut updated_origins = vec![];
+        for sn_uuid in snapshots {
+            if let Some((_, fs)) = self.filesystems.get_mut_by_uuid(sn_uuid) {
+                if fs.unset_origin() {
+                    updated_origins.push(sn_uuid);
+                }
+            }
+        }
```

Comment:
I think this could all be collapsed into a single iterator using `filesystems_mut()` for the iterator and `fold()` at the end to collect the updated origins.